modules/caddyfs/filesystem.go

```diff
@@ -0,0 +1,111 @@
+package caddyfs
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/fs"
+
+	"github.com/caddyserver/caddy/v2"
+	"github.com/caddyserver/caddy/v2/caddyconfig"
+	"github.com/caddyserver/caddy/v2/caddyconfig/caddyfile"
+	"github.com/caddyserver/caddy/v2/caddyconfig/httpcaddyfile"
+	"go.uber.org/zap"
+)
+
+func init() {
+	caddy.RegisterModule(Filesystems{})
+	httpcaddyfile.RegisterGlobalOption("filesystem", parseFilesystems)
+}
+
+type moduleEntry struct {
+	Key           string          `json:"name,omitempty"`
+	FileSystemRaw json.RawMessage `json:"file_system,omitempty" caddy:"namespace=caddy.fs inline_key=backend"`
+	fileSystem    fs.FS
+}
+
+// Filesystems loads caddy.fs modules into the global filesystem map
+type Filesystems struct {
+	Filesystems []*moduleEntry `json:"filesystems"`
+
+	defers []func()
+}
+
+func parseFilesystems(d *caddyfile.Dispenser, existingVal any) (any, error) {
+	p := &Filesystems{}
+	current, ok := existingVal.(*Filesystems)
+	if ok {
+		p = current
+	}
+	x := &moduleEntry{}
+	err := x.UnmarshalCaddyfile(d)
+	if err != nil {
+		return nil, err
+	}
+	p.Filesystems = append(p.Filesystems, x)
+	return p, nil
+}
+
+// CaddyModule returns the Caddy module information.
+func (Filesystems) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID:  "caddy.filesystems",
+		New: func() caddy.Module { return new(Filesystems) },
+	}
+}
+
+func (xs *Filesystems) Start() error { return nil }
+func (xs *Filesystems) Stop() error  { return nil }
+
+func (xs *Filesystems) Provision(ctx caddy.Context) error {
+	// load the filesystem module
+	for _, f := range xs.Filesystems {
+		if len(f.FileSystemRaw) > 0 {
+			mod, err := ctx.LoadModule(f, "FileSystemRaw")
+			if err != nil {
+				return fmt.Errorf("loading file system module: %v", err)
+			}
+			f.fileSystem = mod.(fs.FS)
+		}
+		// register that module
+		ctx.Logger().Debug("registering fs", zap.String("fs", f.Key))
+		ctx.Filesystems().Register(f.Key, f.fileSystem)
+		// remember to unregister the module when we are done
+		xs.defers = append(xs.defers, func() {
+			ctx.Filesystems().Unregister(f.Key)
+		})
```

Comment:
How does this behave with config reloads? Is ctx scoped to only the current config (and not global)? Just making sure that it won't unregister an fs that didn't change keys between config reloads.

Reply:
so I assumed caddy would run cleanup before the config reload and provision after, is that not what happens for config reloads?

Reply:
See https://caddyserver.com/docs/architecture for an explanation of the lifecycle.

Basically, when reloading there's two configs loaded and started for a moment. The new config is loaded and started before the old config is stopped. Cleanup gets called during stoppage.

I think what you did might be fine, I just wanted to make sure you tested config reloads and confirmed things still work normally after a reload.

Reply:
i haven't so i will, and i will test a few cases as well. 

is there a way to programmatically test this with some sort of harness as well?